### DISCUSSION TITLE
Define error type

### DISCUSSION THREAD

**Discussion by kholiavko-roman** [TanStack/query#7177] [Q&A] [answered]
How to set an error type to avoid ts errors like

<img width="495" alt="image" src="https://github.com/TanStack/query/assets/16399378/902a59f3-461d-46cb-a332-d624a3f61656">

All the answers suggest checking the error by type in runtime, but it's not quite useful in some cases
https://stackoverflow.com/questions/72321623/whats-the-type-of-react-querys-error-and-how-to-handle-different-cases

Is there a different way to type an error?

I know that in version 4 it was possible to use `onError` and set an error type in callback parameter, and after that the error that is returned from the `useQuery` has those types. How we can do something like that in v5? 

Thanks.

**Comment by TkDodo** [comment, marked as answer]
https://tanstack.com/query/v5/docs/framework/react/typescript#typing-the-error-field

**Reply by kholiavko-roman** [reply]
Thanks!

**Reply by kevinashworth** [reply]
So when a page like https://tanstack.com/query/latest/docs/framework/react/reference/useQuery lists the error passed to `retry`, `retryDelay` and `throwOnError` as `TError`, it should really be `Error`?

**Reply by TkDodo** [reply]
`TError` is a type parameter that defaults to `Error` (or whatever you have registered)

**Reply by helmax-y** [reply]
@TkDodo and what's the way to set a type for `throwOnError` callback, which is set globally via QueryClient? I want to apply this error handling approach https://tkdodo.eu/blog/react-query-error-handling#error-boundaries, but for `(error) => error.response?.status >= 500` I get TS error `Property 'response' does not exist on type 'Error'`

**Reply by TkDodo** [reply]
you still literally need this:

https://tanstack.com/query/v5/docs/framework/react/typescript#typing-the-error-field

or perform type narrowing with `'response' in error`

**Comment by dinhuty** [comment]
options?: UseMutationOptions<
    ResponseType,
    AxiosError<any>,
    PrameterType,
    unknown
  >
